xds/internal/server/listener_wrapper.go

```diff
@@ -253,36 +353,10 @@ func (l *listenerWrapper) Accept() (net.Conn, error) {
 			conn.Close()
 			continue
 		}
-		var rc xdsresource.RouteConfigUpdate
-		if fc.InlineRouteConfig != nil {
-			rc = *fc.InlineRouteConfig
-		} else {
-			rcPtr := atomic.LoadPointer(&l.rdsUpdates)
-			rcuPtr := (*map[string]xdsresource.RouteConfigUpdate)(rcPtr)
-			// This shouldn't happen, but this error protects against a panic.
-			if rcuPtr == nil {
-				return nil, errors.New("route configuration pointer is nil")
-			}
-			rcu := *rcuPtr
-			rc = rcu[fc.RouteConfigName]
-		}
-		// The filter chain will construct a usuable route table on each
-		// connection accept. This is done because preinstantiating every route
-		// table before it is needed for a connection would potentially lead to
-		// a lot of cpu time and memory allocated for route tables that will
-		// never be used. There was also a thought to cache this configuration,
-		// and reuse it for the next accepted connection. However, this would
-		// lead to a lot of code complexity (RDS Updates for a given route name
-		// can come it at any time), and connections aren't accepted too often,
-		// so this reinstantation of the Route Configuration is an acceptable
-		// tradeoff for simplicity.
-		vhswi, err := fc.ConstructUsableRouteConfiguration(rc)
-		if err != nil {
-			l.logger.Warningf("Failed to construct usable route configuration: %v", err)
-			conn.Close()
-			continue
-		}
-		return &connWrapper{Conn: conn, filterChain: fc, parent: l, virtualHosts: vhswi}, nil
+		cw := &connWrapper{Conn: conn, filterChain: fc, parent: l, rc: fc.RC}
+		l.activeFilterChainManager.AddConn(cw)
```

Comment:
What is the rationale behind persisting the connections in the filter chain manager instead of persisting them here? I don't see the filter chain manager using the conns in anyway. It simply adds the conns to a slice and returns them when asked for.

Also conceptually, it doesn't make sense to me for a filter chain manager to hold the connections instead of the listener wrapper.

Reply:
I had extensive discussions around Eric about this, as this is how Java does this. The reason is the filter chain manager continues to serve until the pending filter chain manager (from a new LDS) goes ready. Thus, it needs to gracefully close any conns when the new one goes ready. Consider the case where you get Conns A, B on first serving fcm, and then switch and gracefully close these conns. Then, you get Conns C and D on the new Conn. This becomes a lot more tricky and nuanced if you persist Conns in the listener wrapper to link to the filter chain manager it was accepted on. The main reason it makes sense in the filter chain manager is you need to gracefully close connections *accepted on that filter chain manager* when it transitions away, so it is tied to the filter chain managers lifecycle. I also think that logically it makes sense that a filter chain manager keeps track of Conn's that receive configuration from the filter chain manager, as the Conns receive configuration based on the filter chain managers configuration, and not on a listener wrappers configuration.